webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/NodeStatusReportRepository.scala

```diff
@@ -160,20 +160,34 @@ class NodeStatusReportRepositoryImpl(
                          // if we are asked to keep compliance, start by checking we do have one
                          case RunAnalysisKind.KeepLastCompliance =>
                            m.get(id) match {
-                             case Some(e) =>
-                               e.runInfo.kind match { // here we need to check for all expiring status from `def outDatedCompliance`
+                             case Some(cached) =>
+                               cached.runInfo.kind match { // here we need to check for all expiring status from `def outDatedCompliance`
                                  case RunAnalysisKind.ComputeCompliance | RunAnalysisKind.Pending =>
                                    // For ComputeCompliance: it's the standard case of keeping existing compliance.
                                    // If the saved node is pending, it means that we are in a case where the node never saved
                                    // anything else than pending since it was first processed, so the node never had a computed
                                    // compliance (it was a new report, directly in pending). Let it blue.
                                    // In both case: keep existing compliance, change run info to avoid loop.
-                                   Some((id, e.modify(_.runInfo).setTo(report.runInfo)))
+
+                                   // we need to copy new report information BUT NOT the one regarding "last run",
+                                   // which were lost if we reach that point
+                                   val runInfo = report.runInfo.copyLastRunInfo(cached.runInfo)
+                                   Some((id, cached.modify(_.runInfo).setTo(runInfo)))
```

Comment:
i'm unsure about this change: I don't see when the received report actually gets to the cache - the runInfo is overriden by the cached.runInfo

Reply:
Ok, no it is actually correct. I'm adding a test to be assured of that.